man/jpeg-zfpoint.1
.TH "jpeg-zfpoint" 1 2.6.4 "14 Feb 2023" "User manual"

.SH NAME
jpeg-zfpoint

.SH DESCRIPTION
Find ZF point of compress JPEGs by re-encoding to the smallest JPEG quality
while keeping perceived visual quality the same and by making sure huffman tables are optimized.
This is a lossy operation, but the images are visually identical and it usually
saves 30-70% of the size for JPEGs coming from a digital camera, particularly DSLRs.
By default all EXIF/IPTC/XMP and color profile metadata is copied over,
but this can be disabled to save more space if desired.

.SH SYNOPSIS
jpeg-zfpoint [options] image.jpg zfpoint.jpg

.SH OPTIONS
.TP
\fB\-a\fR, \fB\-\-accurate\fR
favor accuracy over speed
.TP
\fB\-c\fR, \fB\-\-no-copy\fR
disable copying files that will not be compressed
.TP
\fB\-d\fR, \fB\-\-defish\fR [arg]
set defish strength [0.0]
.TP
\fB\-h\fR, \fB\-\-help\fR
output program help
.TP
\fB\-l\fR, \fB\-\-loops\fR [arg]
set the number of runs to attempt [6]
.TP
\fB\-n\fR, \fB\-\-min\fR [arg]
minimum JPEG quality [40]
.TP
\fB\-p\fR, \fB\-\-no-progressive\fR
disable progressive encoding
.TP
\fB\-r\fR, \fB\-\-ppm\fR
parse input as PPM
.TP
\fB\-s\fR, \fB\-\-strip\fR
strip metadata
.TP
\fB\-x\fR, \fB\-\-max\fR [arg]
maximum JPEG quality [100]
.TP
\fB\-z\fR, \fB\-\-zoom\fR [arg]
set defish zoom [1.0]
.TP
\fB\-A\fR, \fB\-\-radius\fR [arg]
sharpen radius [2]
.TP
\fB\-Q\fR, \fB\-\-quiet\fR
only print out errors
.TP
\fB\-S\fR, \fB\-\-subsample\fR [arg]
set subsampling method, valid values: 'default', 'disable' [default]
.TP
\fB\-T\fR, \fB\-\-input-filetype\fR [arg]
set input file type to one of 'auto', 'jpeg', 'ppm' [auto]
.TP
\fB\-V\fR, \fB\-\-version\fR
output program version
.TP
\fB\-Y\fR, \fB\-\-ycbcr\fR [arg]
YCbCr jpeg colorspace: 0 - source, >0 - YCrCb, <0 - RGB
.SH EXAMPLES
Default settings:
.PP
.I
jpeg-zfpoint image.jpg zfpoint.jpg

.SH COPYRIGHT
 JPEG-Archive is copyright © 2015 Daniel G. Taylor
 Image Quality Assessment (IQA) is copyright 2011, Tom Distler (http://tdistler.com)
 SmallFry is copyright 2014, Derek Buitenhuis (https://github.com/dwbuiten)
 All rights reserved.

.SH "SEE ALSO"
 jpeg-compare,
 jpeg-hash,
 jpeg-recompress,
 webp-compress,
 cjpeg


CHANGELOG
JPEG Recompress
https://github.com/ImageProcessing-ElectronicPublications/jpeg-recompress

2.6.4 "um"

    Research: correct UM model: all metrics.
    2023, zvezdochiot

2.6.3 "um"

    Research: correct UM model.
    2023, zvezdochiot

2.6.2 "sharpenbad"

    reimplementation SHARPENBAD:
      metric_sharpenbad

2.6.1 "msef"

    Added MSE and MSEF metric.
    Research: correct trends UM (44 images)

2.6.0 "sum"

    Added VIFP 1 layer in SUM metric.

2.5.9 "vifp1"

    Added VIFP 1 layer metric.

2.5.8 "neat"

    Neat adds up all values.

2.5.7 "bisection"

    Fix bisection

2.5.6 "subsample"

    Fix IPEP/#1: ignores subsampling option

2.5.5 "sigma"

    Move `sigma`-function from `libsmallfry` to `jmetrics.a`
    Research: correct trends UM (32 images)
    Remove metrics `cor` and `corsh`
    2022, zvezdochiot

2.5.4 "near"

    Research: correct trends UM (24 images)
    2022, zvezdochiot

2.5.3 "approx"

    Research: correct trends UM (16 images)
    2022, zvezdochiot

2.5.2 "UM"

    Research: correct trends UM (8 images)
    2022, zvezdochiot

2.5.1 "saved"

    Fix overflow saved.
    Correct quality scale (next = (1.0 + prev) / 2)
    Replacing multiplications with counters.
    2022, zvezdochiot

2.5.0 "webp"

    Use libwebp for WEBP format
    2022, zvezdochiot

2.4.0 "nhw"

    Update for libsmallfry 0.1.0
    metric_nhw
    2022, zvezdochiot

2.3.1 "weight"

    weight average for SUM metrics
    2021, zvezdochiot

2.3.0 "return"

    return from mozjpeg to jpeg8
    2021, zvezdochiot

2.2.1 "linearize"

    jpeg-recompress: linearize metric
    2018, zvezdochiot

2.2 "zfpoint"

    Find ZF point quality for JPEG
    Update for libsmallfry 0.0.3
    2018, zvezdochiot

2.1.3 "sum"

    jpeg-recompress: summary metric and linearize ssim
    2018, zvezdochiot

2.1.2

    Update for libsmallfry 0.0.2
    sharpenbad_metric
    2018, zvezdochiot

2.1.1

    fork jpeg-archive 2.1.1 to jpeg-recompress for compatible libsmallfry.
    Daniel G. Taylor & zvezdochiot


src/jpeg-zfpoint.c
/*
    Find ZF point of compress a JPEG file while attempting to keep visual quality the same
    by using structural similarity (SSIM) as a metric. Does a binary search
    between JPEG quality 40 and 95 to find the best match. Also makes sure
    that huffman tables are optimized if they weren't already.
*/

#include <getopt.h>
#include "jmetrics.h"

const char *COMMENT = "Find ZF point of compress a JPEG";

void usage(char *progname)
{
    printf("usage: %s [options] input.jpg output.jpg\n\n", progname);
    printf("options:\n\n");
    printf("  -a, --accurate               favor accuracy over speed\n");
    printf("  -c, --no-copy                disable copying files that will not be compressed\n");
    printf("  -d, --defish [arg]           set defish strength [0.0]\n");
    printf("  -f, --force                  force process\n");
    printf("  -h, --help                   output program help\n");
    printf("  -l, --loops [arg]            set the number of runs to attempt [6]\n");
    printf("  -n, --min [arg]              minimum JPEG quality [40]\n");
    printf("  -p, --no-progressive         disable progressive encoding\n");
    printf("  -r, --ppm                    parse input as PPM\n");
    printf("  -s, --strip                  strip metadata\n");
    printf("  -x, --max [arg]              maximum JPEG quality [100]\n");
    printf("  -z, --zoom [arg]             set defish zoom [1.0]\n");
    printf("  -A, --radius [arg]           sharpen radius [2]\n");
    printf("  -Q, --quiet                  only print out errors\n");
    printf("  -S, --subsample [arg]        set subsampling method to one of 'default', 'disable' [default]\n");
    printf("  -T, --input-filetype [arg]   set input file type to one of 'auto', 'jpeg', 'ppm' [auto]\n");
    printf("  -V, --version                output program version\n");
    printf("  -Y, --ycbcr [arg]            YCbCr jpeg colorspace: 0 - source, >0 - YCrCb, <0 - RGB\n");
}

int main (int argc, char **argv)
{
    // Number of binary search steps
    int attempts = 6;

    // Min/max JPEG quality
    int jpegMin = 40;
    int jpegMax = 100;
    int shRadius = 2;

    int force = 0;
    int ycbcr = 0;
    // Strip metadata from the file?
    int strip = 0;

    // Disable progressive mode?
    int noProgressive = 0;

    // Defish the image?
    float defishStrength = 0.0;
    float defishZoom = 1.0;

    // Input format
    enum filetype inputFiletype = FILETYPE_AUTO;

    // Whether to copy files that cannot be compressed
    int copyFiles = 1;

    // Whether to favor accuracy over speed
    int accurate = 0;

    // Chroma subsampling method
    int subsample = SUBSAMPLE_DEFAULT;

    // Quiet mode (less output)
    int quiet = 0;

    unsigned char *buf, *original, *originalGray = NULL, *tmpImage;
    unsigned char *compressed = NULL, *compressedGray, *metaBuf;
    long bufSize = 0, originalSize = 0, originalGraySize = 0;
    long compressedGraySize = 0;
    unsigned long compressedSize = 0, saved;
    int width, height, min, max, attempt, percent, app0_len;
    int jpegcs, jpegcst, quality, progressive, optimize;
    unsigned int metaSize = 0;
    float metric, maxmetric, qmetric, cmpMin, cmpMax, cmpQ;
    char *inputPath, *outputPath;
    FILE *file;

    const char *optstring = "acd:fhl:n:prsx:z:A:S:T:QVY:";
    static const struct option opts[] =
    {
        { "accurate", no_argument, 0, 'a' },
        { "defish", required_argument, 0, 'd' },
        { "force", no_argument, 0, 'f' },
        { "help", no_argument, 0, 'h' },
        { "input-filetype", required_argument, 0, 'T' },
        { "loops", required_argument, 0, 'l' },
        { "min", required_argument, 0, 'n' },
        { "max", required_argument, 0, 'x' },
        { "no-copy", no_argument, 0, 'c' },
        { "no-progressive", no_argument, 0, 'p' },
        { "ppm", no_argument, 0, 'r' },
        { "quiet", no_argument, 0, 'Q' },
        { "radius", required_argument, 0, 'A' },
        { "strip", no_argument, 0, 's' },
        { "subsample", required_argument, 0, 'S' },
        { "version", no_argument, 0, 'V' },
        { "ycbcr", required_argument, 0, 'Y' },
        { "zoom", required_argument, 0, 'z' },
        { 0, 0, 0, 0 }
    };
    int opt, longind = 0;

    char *progname = "jpeg-zfpoint";

    while ((opt = getopt_long(argc, argv, optstring, opts, &longind)) != -1)
    {
        switch (opt)
        {
        case 'a':
            accurate = 1;
            break;
        case 'c':
            copyFiles = 0;
            break;
        case 'd':
            defishStrength = atof(optarg);
            break;
        case 'f':
            force = 1;
            break;
        case 'h':
            usage(progname);
            return 0;
        case 'l':
            attempts = atoi(optarg);
            break;
        case 'n':
            jpegMin = atoi(optarg);
            break;
        case 'p':
            noProgressive = 1;
            break;
        case 'r':
            inputFiletype = FILETYPE_PPM;
            break;
        case 's':
            strip = 1;
            break;
        case 'x':
            jpegMax = atoi(optarg);
            break;
        case 'z':
            defishZoom = atof(optarg);
            break;
        case 'A':
            shRadius = atof(optarg);
            break;
        case 'Q':
            quiet = 1;
            break;
        case 'S':
            subsample = parseSubsampling(optarg);
            break;
        case 'T':
            if (inputFiletype != FILETYPE_AUTO)
            {
                error("multiple file types specified for the input file");
                return 1;
            }
            inputFiletype = parseInputFiletype(optarg);
            break;
        case 'V':
            version();
            return 0;
        case 'Y':
            ycbcr = atoi(optarg);
            break;
        };
    }

    if (argc - optind != 2)
    {
        usage(progname);
        return 255;
    }

    inputPath = argv[optind];
    outputPath = argv[optind + 1];

    /* Read the input into a buffer. */
    bufSize = readFile(inputPath, (void **) &buf);

    /* Detect input file type. */
    if (inputFiletype == FILETYPE_AUTO)
        inputFiletype = detectFiletypeFromBuffer(buf, bufSize);

    /*
     * Read original image and decode. We need the raw buffer contents and its
     * size to obtain meta data and the original file size later.
     */
    originalSize = decodeFileFromBuffer(buf, bufSize, &original, inputFiletype, &width, &height, &jpegcs, JCS_RGB);
    if (!originalSize)
    {
        error("invalid input file: %s", inputPath);
        return 1;
    }

    if (defishStrength)
    {
        info(quiet, "Defishing...\n");
        tmpImage = malloc(width * height * 3);
        defish(original, tmpImage, width, height, 3, defishStrength, defishZoom);
        free(original);
        original = tmpImage;
    }

    // Convert RGB input into Y
    originalGraySize = grayscale(original, &originalGray, width, height);

    if (inputFiletype == FILETYPE_JPEG)
    {
        // Read metadata (EXIF / IPTC / XMP tags)
        if (getMetadata(buf, bufSize, &metaBuf, &metaSize, COMMENT) && !force)
        {
            if (copyFiles)
            {
                info(quiet, "File already processed by jpeg-zfpoint!\n");
                file = openOutput(outputPath);
                if (file == NULL)
                {
                    error("could not open output file: %s", outputPath);
                    return 1;
                }

                fwrite(buf, bufSize, 1, file);
                fclose(file);

                free(buf);

                return 0;
            }
            else
            {
                error("file already processed by jpeg-zfpoint!");
                free(buf);
                return 2;
            }
        }
    }

    if (strip)
        metaSize = 0;
    else
        info(quiet, "Metadata size is %ukb\n", metaSize / 1024);

    if (!originalSize || !originalGraySize)
        return 1;

    if (ycbcr < 0)
        jpegcs = JCS_RGB;
    if (ycbcr > 0)
        jpegcs = JCS_YCbCr;

    if (jpegMin > jpegMax)
    {
        error("maximum JPEG quality must not be smaller than minimum JPEG quality!");
        return 1;
    }

    // Find ZF point.
    min = jpegMin;
    max = jpegMax;
    compressedSize = encodeJpeg(&compressed, original, width, height, JCS_RGB, max, jpegcs, 0, 1, subsample);
    compressedGraySize = decodeJpeg(compressed, compressedSize, &compressedGray, &width, &height, &jpegcst, JCS_GRAYSCALE);
    maxmetric = metric_corsharp(originalGray, compressedGray, width, height, shRadius);
    maxmetric = MetricSigma(maxmetric);
    qmetric = maxmetric / (float)max;
    cmpMax = 0;
    compressedSize = encodeJpeg(&compressed, original, width, height, JCS_RGB, min, jpegcs, 0, 1, subsample);
    compressedGraySize = decodeJpeg(compressed, compressedSize, &compressedGray, &width, &height, &jpegcst, JCS_GRAYSCALE);
    metric = metric_corsharp(originalGray, compressedGray, width, height, shRadius);
    metric = MetricSigma(metric);
    cmpMin = qmetric * (float)min - metric;
    for (attempt = attempts - 1; attempt >= 0; --attempt)
    {
        quality = (max + min + 1) / 2;
        progressive = attempt ? 0 : !noProgressive;
        optimize = accurate ? 1 : (attempt ? 0 : 1);

        // Recompress to a new quality level, without optimizations (for speed)
        compressedSize = encodeJpeg(&compressed, original, width, height, JCS_RGB, quality, jpegcs, progressive, optimize, subsample);

        // Load compressed luma for quality comparison
        compressedGraySize = decodeJpeg(compressed, compressedSize, &compressedGray, &width, &height, &jpegcst, JCS_GRAYSCALE);

        if (!compressedGraySize)
        {
            fprintf(stderr, "Unable to decode file that was just encoded!\n");
            return 1;
        }

        if (!attempt)
        {
            info(quiet, "Final optimized ");
        }

        // Measure quality difference
        metric = metric_corsharp(originalGray, compressedGray, width, height, shRadius);
        metric = MetricSigma(metric);
        cmpQ = qmetric * (float)quality - metric;
        info(quiet, "zfpoint");

        if (attempt)
            info(quiet, " at q=%i (%i - %i): dM %f\n", quality, min, max, cmpQ);
        else
            info(quiet, " at q=%i: dM %f\n", quality, cmpQ);

        if (cmpMin < cmpMax)
        {
            min = MIN(quality, max);
            cmpMin = cmpQ;
        }
        else
        {
            max = MAX(quality, min);
            cmpMax = cmpQ;
        }

        // If we aren't done yet, then free the image data
        if (attempt)
        {
            free(compressed);
            free(compressedGray);
        }
    }

    free(buf);


    // Calculate and show savings, if any
    percent = (compressedSize + metaSize) * 100 / bufSize;
    saved = (bufSize > (compressedSize + metaSize)) ? (bufSize - compressedSize - metaSize) : 0;
    info(quiet, "New size is %i%% of original (saved %lu kb)\n", percent, saved / 1024);

    if (compressedSize >= bufSize && !force)
    {
        error("output file is larger than input, aborting!");
        return 1;
    }

    // Open output file for writing
    file = openOutput(outputPath);
    if (file == NULL)
    {
        error("could not open output file");
        return 1;
    }

    /* Check that the metadata starts with a SOI marker. */
    if (!checkJpegMagic(compressed, compressedSize))
    {
        error("missing SOI marker, aborting!");
        return 1;
    }

    /* Make sure APP0 is recorded immediately after the SOI marker. */
    if (compressed[2] != 0xff || (compressed[3] != 0xe0 && compressed[3] != 0xee))
    {
        error("missing APP0 marker, aborting!");
        return 1;
    }

    /* Write SOI marker and APP0 metadata to the output file. */
    app0_len = (compressed[4] << 8) + compressed[5];
    fwrite(compressed, 4 + app0_len, 1, file);

    /*
     * Write comment (COM metadata) so we know not to reprocess this file in
     * the future if it gets passed in again.
     */
    fputc(0xff, file);
    fputc(0xfe, file);
    fputc(0x00, file);
    fputc(strlen(COMMENT) + 2, file);
    fwrite(COMMENT, strlen(COMMENT), 1, file);

    /* Write additional metadata markers. */
    if (inputFiletype == FILETYPE_JPEG && !strip)
        fwrite(metaBuf, metaSize, 1, file);

    /* Write image data. */
    fwrite(compressed + 4 + app0_len, compressedSize - 4 - app0_len, 1, file);
    fclose(file);

    if (inputFiletype == FILETYPE_JPEG && !strip)
        free(metaBuf);

    free(compressed);
    free(original);
    free(originalGray);

    return 0;
}
